text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+(while the implementation of `libpanic_unwind` itself may change
+to no longer maintain that compatibility).
+
+In order for Cargo users to be able
+to specify the C++ compatible `panic_runtime` implementation,
+a new optional value, runtime, will be added to the `profile.foo.panic` option:
+
+```toml
+[profile.dev]
+panic = { 'unwind', runtime = 'native' }
+```
+
+`panic.runtime` may only be specified if the panic strategy is `unwind`.
+And just as there may only be one strategy
+in the dependency graph for a final product,
+there may only be one `runtime`.
+
+For now, the only valid `runtime` keys will be:
+
+* `default` - identical to `panic = 'unwind'` with no `runtime` selected.
+  This will use Rust's default unwinding runtime,
+  unless another crate in the final product's dependency graph
+specifies `panic = 'abort'`,
+  in which case the `abort` strategy will take precedence as usual.
+  which is not guaranteed to be compatible with native exceptions.
+* `native` - as discussed above, this will preserve the behavior of the current
+  implementation of `libpanic_unwind`.
+* `crate` - indicates that a non-`std` crate will use `#![panic_runtime]` to
+  provide the runtime.
+* `self` - indicates that this crate itself provides `#![panic_runtime]`.
+
+If the `native` or `crate` key is specified
+anywhere in a final product's dependency graph,
+no crate in that dependency graph may specify the `panic = abort` strategy;
+this mismatch will cause the build to fail.
+
+The function annotation `#[panic(allowed)]` will only be permitted in crates
+that specify a non-default `panic.runtime`.
+
+Crates that do not use the `profile.foo.panic` option at all
+will remain compatible with any `profile.foo.panic` configuration
+used to generate the final product.
+
+For non-Cargo users, equivalent `rustc` flags will be provided
+(which will be how Cargo itself implements the option).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Unwinding for functions with the `#[unwind(allowed)]` annotation
+is performed as if the function were not marked `extern`.
+This annotation has no effect on functions not marked `extern`.
+It has no observable effect unless the marked function `panic`s
+(e.g. it has no observable effect
+when a function returns normally or enters an infinite loop).
+The LLVM IR for such functions must not be marked `noexcept`.
+
+The compiler will have a new stable flag, `-C panic.runtime`,
+which will be required to enable the `#[unwind(allowed)]` annotation;
+as explained above, the flag will specify the expected source
+of the panic runtime to be used in the final (binary) product.
+If the source is `default`,
+the `libpanic_unwind` crate will provide the runtime,
+and the `#[unwind(allowed)]` annotation will not be permitted in that crate.
+
+Different values of `panic.runtime` will not be permitted for different crates
+in a single dependency graph,
```

Comment:
What happens if two objects are not part of the same dependency graph, but, e.g., are dynamically linked ? E.g. some Rust binary is dynamically linked against a Rust dynamic library that exposes a `#[unwind(allowed)] extern "Rust" fn foo() {}` function.

Reply:
As far as I know, this is entirely out of our control, as I mentioned above. (The decision to link or not is essentially in the hands of the target platforms binary loader, right?) That's part of why the annotation needs `unsafe`.

Reply:
The "right" answer here, I think, is to define that the runtime is part of the ABI of the function. Allowing the function declaration to define what panic runtime it's using would solve the dynamic linking problem safely, at the cost of forcing us to stabilize the details of the ABI. For `panic.runtime = "native"`, those details are already forced onto us anyways, but I'm not sure we want to force ourselves to take the step for future runtimes as well.

Reply:
@jcranmer I may be misunderstanding you, but individual functions within the same shared library can't have different ABIs, can they? What would it mean for the runtime to be "part of the ABI of the function"?

Reply:
> but individual functions within the same shared library can't have different ABIs, can they?

Of course, `extern "C"`, `extern "fastcall"`, `extern "Rust`, etc. 

> What would it mean for the runtime to be "part of the ABI of the function"?

Exactly that. In the same way one has to declare a function `extern "C"`  when calling a function following that ABI, and using some other `extern "X"` when that function uses a different ABI, one would need to specify what the panic ABI of a function that can panic is. 

That's what `#[unwind(Rust)]` did. It specified that this function can panic "with the Rust ABI". Since there can only be one panic implementation for all dynamically and statically linked Rust libraries, `#[unwind(Rust)]` meant "whatever ABI that has".

Reply:
Can system linkers actually detect ABI mismatches? How is this done?

Reply:
> Can system linkers actually detect ABI mismatches? 

No. This is not a new problem, if you declare a function as `extern "C"`, and that function happens to have a different ABI, you often don't get linker errors either.